Tests/Controller/AzineEmailControllerTest.php

```diff
@@ -0,0 +1,177 @@
+    {
+        $this->checkApplication();
+
+        // Create a new client to browse the application
+        $client = static::createClient();
+        $client->followRedirects();
+
+        $manager = $this->getEntityManager();
+        $sentEmailRep = $manager->getRepository("Azine\EmailBundle\Entity\SentEmail");
+
+        $testSentEmails = $sentEmailRep->search(['recipients' => TestHelper::TEST_EMAIL]);
+
+
+        if (count($testSentEmails) == 0) {
+
+            TestHelper::addSentEmails($manager);
+        }
+
+        $listUrl = substr($this->getRouter()->generate("azine_admin_email_dashboard", array('_locale' => "en")), 13);
+        $crawler = $this->loginUserIfRequired($client, $listUrl);
+
+        //click on an email web view link to get to the web page
+        $link = $crawler->filter(".sentEmail:contains('".EmailTemplateProvider::NEWSLETTER_TEMPLATE."')")->first()->filter("td")->last()->filter("a")->first()->link();
+        $crawler = $client->click($link);
+
+        $this->assertEquals(1, $crawler->filter("span:contains('_az.email.hello')")->count(), " div with hello message expected.");
+
+        $crawler = $this->loginUserIfRequired($client, $listUrl);
+
+        //Test filtering by email
+        $crawler = $crawler->selectButton('sentEmail[save]');
+        $form = $crawler->form();
+        $form['sentEmail[recipients]'] = TestHelper::TEST_EMAIL;
+        $crawler = $client->submit($form);
+
+        $this->assertEquals($crawler->filter(".sentEmail")->count(), $crawler->filter("tr:contains('".TestHelper::TEST_EMAIL."')")->count(),"Table rows only with ".TestHelper::TEST_EMAIL." email are expected");
+
+
+        //click on email details view link to get to the details page
+        $link = $crawler->filter(".sentEmail:contains('".TestHelper::TEST_EMAIL."')")->first()->filter("td")->last()->filter("a")->last()->link();
+        $crawler = $client->click($link);
+
+        $this->assertEquals(1, $crawler->filter("tr:contains('".TestHelper::TEST_EMAIL."')")->count(),"Table cell with ".TestHelper::TEST_EMAIL." expected");
+
+        $crawler = $this->loginUserIfRequired($client, $listUrl);
+
+        $form['sentEmail[recipients]'] = '';
+        $crawler = $client->submit($form);
+
+        //Test filtering by token
+        $form['sentEmail[token]'] = TestHelper::TEST_TOKEN;
+        $crawler = $client->submit($form);
+
+        $this->assertEquals($crawler->filter(".sentEmail")->count(), $crawler->filter(".sentEmail:contains('".TestHelper::TEST_TOKEN."')")->count(),"Table row only with ".TestHelper::TEST_TOKEN." token is expected");
+
+    }
+
+    /**
+     * Load the url and login if required.
+     * @param  string  $url
+     * @param  string  $username
+     * @param  Client  $client
+     * @return Crawler $crawler of the page of the url or the page after the login
+     */
+    private function loginUserIfRequired(Client $client, $url, $username = "dominik", $password = "lkjlkjlkjlkj")
```

Comment:
Loging in as admin should be done once explicitly. The following calls to a URL should not go through the login-logic everytime. => Then  `$crawler = $client->request("GET", $url); ` can be called instead of `$crawler = $this->loginUserIfRequired(...`